File: examples/example01.cpp
/****************************************************************************
 *
 * Copyright 2026 Thiago Marques de Oliveira
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 ****************************************************************************/

/**
 * @brief Example state machine node: engage Offboard mode then arm the vehicle
 * @file offboard_arm_example.cpp
 *
 * State flow:
 *
 *   IDLE ──► ENGAGE_OFFBOARD ──► ARM ──► FLYING ──► ERROR (on any failure)
 *
 * @author Thiago Marques de Oliveira
 * @date April, 2026
 */


/* Private includes ------------------------------------------------------ */

#include "px4_commander/px4_commander_client.hpp"

#include <rclcpp/rclcpp.hpp>

#include <chrono>
#include <functional>
#include <string>


/* Namespaces ------------------------------------------------------------ */

using namespace std::chrono_literals;


/* OffboardArmNode class ------------------------------------------------- */

class OffboardArmNode : public rclcpp::Node
{
public:

    // --------------------------------------------
    //   STATE MACHINE
    // --------------------------------------------

    enum class State
    {
        IDLE,
        ENGAGE_OFFBOARD,
        ARM,
        FLYING,
        ERROR
    };


    // --------------------------------------------
    //   CONSTRUCTOR
    // --------------------------------------------

    OffboardArmNode()
    : Node("offboard_arm_example_node"),
      state_(State::IDLE)
    {
        commander_ = std::make_shared<Px4CommanderClient>(
            this->shared_from_this()
        );

        // Tick the state machine at 1 Hz
        timer_ = this->create_wall_timer(
            1s,
            std::bind(&OffboardArmNode::tick, this)
        );

        RCLCPP_INFO(this->get_logger(), "Node started — initial state: IDLE");
    }


private:

    // --------------------------------------------
    //   STATE MACHINE TICK
    // --------------------------------------------

    /**
     * @brief Advance the state machine by one step
     */
    void tick()
    {
        switch (state_)
        {
            case State::IDLE:
                on_idle();
                break;

            case State::ENGAGE_OFFBOARD:
                on_engage_offboard();
                break;

            case State::ARM:
                on_arm();
                break;

            case State::FLYING:
                on_flying();
                break;

            case State::ERROR:
                on_error();
                break;
        }
    }


    // --------------------------------------------
    //   STATE HANDLERS
    // --------------------------------------------

    /**
     * @brief IDLE — entry point, immediately transitions to ENGAGE_OFFBOARD
     */
    void on_idle()
    {
        RCLCPP_INFO(this->get_logger(), "[IDLE] Starting sequence...");
        transition_to(State::ENGAGE_OFFBOARD);
    }

    /**
     * @brief ENGAGE_OFFBOARD — send offboard mode command and wait for confirmation
     */
    void on_engage_offboard()
    {
        RCLCPP_INFO(this->get_logger(), "[ENGAGE_OFFBOARD] Engaging offboard mode...");

        const auto [success, message] = commander_->engage_offboard_mode();

        if (success)
        {
            RCLCPP_INFO(this->get_logger(), "[ENGAGE_OFFBOARD] %s", message.c_str());
            transition_to(State::ARM);
        }
        else
        {
            RCLCPP_ERROR(this->get_logger(), "[ENGAGE_OFFBOARD] %s", message.c_str());
            transition_to(State::ERROR);
        }
    }

    /**
     * @brief ARM — arm the vehicle after offboard mode is confirmed
     */
    void on_arm()
    {
        RCLCPP_INFO(this->get_logger(), "[ARM] Arming vehicle...");

        const auto [success, message] = commander_->arm();

        if (success)
        {
            RCLCPP_INFO(this->get_logger(), "[ARM] %s", message.c_str());
            transition_to(State::FLYING);
        }
        else
        {
            RCLCPP_ERROR(this->get_logger(), "[ARM] %s", message.c_str());
            transition_to(State::ERROR);
        }
    }

    /**
     * @brief FLYING — vehicle is armed and in offboard mode, ready for control
     */
    void on_flying()
    {
        RCLCPP_INFO(this->get_logger(), "[FLYING] Vehicle is flying in offboard mode");

        // Stop the tick timer — nothing else to do in this example
        timer_->cancel();
    }

    /**
     * @brief ERROR — unrecoverable failure, shut down the node
     */
    void on_error()
    {
        RCLCPP_ERROR(this->get_logger(), "[ERROR] Sequence failed — shutting down");
        timer_->cancel();
        rclcpp::shutdown();
    }


    // --------------------------------------------
    //   HELPERS
    // --------------------------------------------

    /**
     * @brief Log and perform a state transition
     * @param next  Target state
     */
    void transition_to(State next)
    {
        RCLCPP_INFO(
            this->get_logger(),
            "State transition: %s → %s",
            state_to_string(state_),
            state_to_string(next)
        );

        state_ = next;
    }

    /**
     * @brief Convert a State enum value to a human-readable string
     * @param state  State to convert
     * @return const char* with the state name
     */
    static const char* state_to_string(State state)
    {
        switch (state)
        {
            case State::IDLE:             return "IDLE";
            case State::ENGAGE_OFFBOARD:  return "ENGAGE_OFFBOARD";
            case State::ARM:              return "ARM";
            case State::FLYING:           return "FLYING";
            case State::ERROR:            return "ERROR";
            default:                      return "UNKNOWN";
        }
    }


    // --------------------------------------------
    //   PRIVATE MEMBERS
    // --------------------------------------------

    State                                state_;
    std::shared_ptr<Px4CommanderClient>  commander_;
    rclcpp::TimerBase::SharedPtr         timer_;
};


// --------------------------------------------
//   NODE INITIALIZATION
// --------------------------------------------

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<OffboardArmNode>();

    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}

File: examples/python/example01.py
#!/usr/bin/env python3

"""
@brief Example state machine node: engage Offboard mode then arm the vehicle
@file example01.py

State flow:

  IDLE ──► ENGAGE_OFFBOARD ──► ARM ──► FLYING ──► ERROR (on any failure)

@author Thiago Marques de Oliveira
@date April, 2026
"""


# Private imports ---------------------------------------------------------

import rclpy
from rclpy.node import Node

from enum import Enum, auto

from px4_commander.px4_commander_client import Px4CommanderClient


# OffboardArmNode class ---------------------------------------------------

class OffboardArmNode(Node):

    # --------------------------------------------
    #   STATE MACHINE
    # --------------------------------------------

    class State(Enum):
        IDLE             = auto()
        ENGAGE_OFFBOARD  = auto()
        ARM              = auto()
        FLYING           = auto()
        ERROR            = auto()


    # --------------------------------------------
    #   CONSTRUCTOR
    # --------------------------------------------

    def __init__(self) -> None:
        super().__init__("offboard_arm_example_node")

        self._state     = self.State.IDLE
        self._commander = Px4CommanderClient()

        # Tick the state machine at 1 Hz
        self._timer = self.create_timer(1.0, self._tick)

        self.get_logger().info("Node started — initial state: IDLE")


    # --------------------------------------------
    #   STATE MACHINE TICK
    # --------------------------------------------

    def _tick(self) -> None:
        """Advance the state machine by one step."""

        dispatch = {
            self.State.IDLE:            self._on_idle,
            self.State.ENGAGE_OFFBOARD: self._on_engage_offboard,
            self.State.ARM:             self._on_arm,
            self.State.FLYING:          self._on_flying,
            self.State.ERROR:           self._on_error,
        }

        dispatch[self._state]()


    # --------------------------------------------
    #   STATE HANDLERS
    # --------------------------------------------

    def _on_idle(self) -> None:
        """IDLE — entry point, immediately transitions to ENGAGE_OFFBOARD."""
        self.get_logger().info("[IDLE] Starting sequence...")
        self._transition_to(self.State.ENGAGE_OFFBOARD)

    def _on_engage_offboard(self) -> None:
        """ENGAGE_OFFBOARD — send offboard mode command and wait for confirmation."""
        self.get_logger().info("[ENGAGE_OFFBOARD] Engaging offboard mode...")

        success, message = self._commander.engage_offboard_mode()

        if success:
            self.get_logger().info(f"[ENGAGE_OFFBOARD] {message}")
            self._transition_to(self.State.ARM)
        else:
            self.get_logger().error(f"[ENGAGE_OFFBOARD] {message}")
            self._transition_to(self.State.ERROR)

    def _on_arm(self) -> None:
        """ARM — arm the vehicle after offboard mode is confirmed."""
        self.get_logger().info("[ARM] Arming vehicle...")

        success, message = self._commander.arm()

        if success:
            self.get_logger().info(f"[ARM] {message}")
            self._transition_to(self.State.FLYING)
        else:
            self.get_logger().error(f"[ARM] {message}")
            self._transition_to(self.State.ERROR)

    def _on_flying(self) -> None:
        """FLYING — vehicle is armed and in offboard mode, ready for control."""
        self.get_logger().info("[FLYING] Vehicle is flying in offboard mode")

        # Stop the tick timer — nothing else to do in this example
        self._timer.cancel()

    def _on_error(self) -> None:
        """ERROR — unrecoverable failure, shut down the node."""
        self.get_logger().error("[ERROR] Sequence failed — shutting down")
        self._timer.cancel()
        rclpy.shutdown()


    # --------------------------------------------
    #   HELPERS
    # --------------------------------------------

    def _transition_to(self, next_state: "OffboardArmNode.State") -> None:
        """Log and perform a state transition."""
        self.get_logger().info(
            f"State transition: {self._state.name} -> {next_state.name}"
        )
        self._state = next_state


# Node initialization -----------------------------------------------------

def main(args=None) -> None:
    rclpy.init(args=args)

    node = OffboardArmNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node._commander.destroy()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()